### PR TITLE
Error on empty dockerfile

### PR DESCRIPTION
**Pull request by LinuxMercedes**
**- What I did**
Fixed a panic when running `docker build` on a dockerfile containing only a comment.

(Docker used to error on this a few years back, but sometime in between now and then something changed.)

**- How I did it**
Detect whether the root AST node is created from a dockerfile with no AST constructs.

**- How to verify it**
```
echo "#hello" > Dockerfile
docker build .
sudo systemctl status docker
```

**- Description for the changelog**

Fixed panic when building dockerfiles containing only comments.

**- A picture of a cute animal (not mandatory but encouraged)**
it's hank, she's my friend

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks! Looks like you forgot to sign-off the commits; you can use the instructions that @GordonTheTurtle (our bot) provided above to fix that

File: builder/remotecontext/detect.go [hunk 1]
```diff
@@ -150,7 +150,17 @@ func readAndParseDockerfile(name string, rc io.Reader) (*parser.Result, error) {
...
 		return nil, errors.Wrap(err, "unexpected error reading Dockerfile")
 	}
-	return parser.Parse(br)
+
+	dockerfile, err := parser.Parse(br)
+	if err != nil {
+		return nil, err
+	}
+
+	if dockerfile.AST == nil || dockerfile.AST.StartLine < 0 {
```
Comment (thaJeztah) [inline builder/remotecontext/detect.go]: Actually; wondering if this check should be done inside `parser.Parse()`. Looks like this check will also be needed in this part of the code: https://github.com/moby/moby/blob/dffc96616503936fc282c0e3a3eea8695ee11b67/builder/remotecontext/detect.go#L34-L39

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks for finding this issue! I left some comments below; I think this fix may actually have to be made in the upstream buildkit repo, where `parse.Parse()` is vendored from; https://github.com/moby/buildkit/blob/3f8ab160d5079539f9ed971fb069e4205108dd9d/frontend/dockerfile/parser/parser.go#L214-L283

File: builder/remotecontext/detect.go (hunk 1 above)
Comment (LinuxMercedes) [reply builder/remotecontext/detect.go]: I thought that might be the place to put it but it seems like an empty/comment-only dockerfile IS a dockerfile, just not one docker cares to do anything with. However, I'm happy to take the patch there instead if that seems like a better place for it!

Comment (thaJeztah) [reply builder/remotecontext/detect.go]: He just replied; and yes, fixing it upstream would be better 👍

Comment (LinuxMercedes) [reply builder/remotecontext/detect.go]: Alright! Should I leave the tests here?

Comment (tonistiigi) [reply builder/remotecontext/detect.go]: Yes, let's reuse these test when updating vendoring. In upstream you can add a unit test for the parser package (and another regression integration test for buildkit itself if you want).

**Comment by tonistiigi** [comment]
> One nice thing about handling this error in moby is that we can include the Dockerfile name if it exists. I feel like that'd be a useful attribute to keep -- do y'all have suggestions on how to reformat the error here to include the filename? Or should I not worry about it?

```
errors.Wrapf(err, "failed to parse %s", name)
```

> Should I update the vendoring in this PR once that PR is merged?

yes

**Comment by LinuxMercedes** [comment]
Alright, I think this is all fixed and ready to go!

**Review by AkihiroSuda — APPROVED** [review]

**Review by fntlnz — APPROVED** [review]
LGTM thanks @LinuxMercedes

**Comment by thaJeztah** [comment]
Thanks! I built this PR and gave it a try;

### 1. With Docker CLI 17.06 and no buildkit

### 1.1 Comment-only Dockerfile:

```bash
docker build -t foo -<<'EOF'
# only a comment
EOF
```

```
Sending build context to Docker daemon  2.048kB
Error response from daemon: failed to parse Dockerfile: file with no instructions.
```

### 1.2 Empty lines only:

```bash
docker build -t foo -<<'EOF'



EOF
```

```
Sending build context to Docker daemon  2.048kB
Error response from daemon: failed to parse Dockerfile: file with no instructions.
```

### 1.3 Empty Dockerfile;

```bash
docker build -t foo -<<'EOF'
EOF
```

```
Sending build context to Docker daemon  1.536kB
Error response from daemon: the Dockerfile (Dockerfile) cannot be empty
```


However, I see we need to make some changes so that the API returns a proper (4xx) error, because it currently doesn't return a typed error, thus a 500 error is returned (see details below);

<details>

```
ERRO[2019-01-07T15:41:02.274166785Z] Handler for POST /v1.30/build returned error: failed to parse Dockerfile: file with no instructions. 
DEBU[2019-01-07T15:41:02.274501230Z] FIXME: Got an API for which error does not match any expected type!!!: file with no instructions.
github.com/docker/docker/vendor/github.com/moby/buildkit/frontend/dockerfile/parser.Parse
	/go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/frontend/dockerfile/parser/parser.go:280
github.com/docker/docker/builder/remotecontext.readAndParseDockerfile
	/go/src/github.com/docker/docker/builder/remotecontext/detect.go:155
github.com/docker/docker/builder/remotecontext.withDockerfileFromContext
	/go/src/github.com/docker/docker/builder/remotecontext/detect.go:77
github.com/docker/docker/builder/remotecontext.newArchiveRemote
	/go/src/github.com/docker/docker/builder/remotecontext/detect.go:58
github.com/docker/docker/builder/remotecontext.Detect
	/go/src/github.com/docker/docker/builder/remotecontext/detect.go:33
github.com/docker/docker/builder/dockerfile.(*BuildManager).Build
	/go/src/github.com/docker/docker/builder/dockerfile/builder.go:88
github.com/docker/docker/api/server/backend/build.(*Backend).Build
	/go/src/github.com/docker/docker/api/server/backend/build/backend.go:60
github.com/docker/docker/api/server/router/build.(*buildRouter).postBuild
	/go/src/github.com/docker/docker/api/server/router/build/build_routes.go:270
github.com/docker/docker/api/server/router/build.(*buildRouter).postBuild-fm
	/go/src/github.com/docker/docker/api/server/router/build/build.go:34
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/experimental.go:26
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/version.go:62
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1
	/go/src/github.com/docker/docker/pkg/authorization/middleware.go:59
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:25
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1
	/go/src/github.com/docker/docker/api/server/server.go:141
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:1964
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
	/go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:162
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2741
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1847
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1333  error_type="*errors.fundamental" module=api
```

</details>


### 2. With a Docker 18.09 cli, and `DOCKER_BUILDKIT=1`:


### 2.1 Comment-only Dockerfile:

```bash
docker build -t foo -<<'EOF'
# only a comment
EOF
```

```
[+] Building 0.1s (2/2) FINISHED                                                                 
 => [internal] load build definition from Dockerfile                                        0.0s
 => => transferring dockerfile: 40B                                                         0.0s
 => [internal] load .dockerignore                                                           0.0s
 => => transferring context: 2B                                                             0.0s
failed to create LLB definition: file with no instructions.
```

### 2.2 Empty lines only:

```bash
docker build -t foo -<<'EOF'



EOF
```
```
[+] Building 0.1s (2/2) FINISHED                                                                 
 => [internal] load build definition from Dockerfile                                        0.0s
 => => transferring dockerfile: 40B                                                         0.0s
 => [internal] load .dockerignore                                                           0.0s
 => => transferring context: 2B                                                             0.0s
failed to create LLB definition: file with no instructions.
```

### 2.3 Empty Dockerfile;

```bash
docker build -t foo -<<'EOF'
EOF
```

```
[+] Building 0.1s (2/2) FINISHED                                                                 
 => [internal] load .dockerignore                                                           0.0s
 => => transferring context: 2B                                                             0.0s
 => [internal] load build definition from Dockerfile                                        0.0s
 => => transferring dockerfile: 31B                                                         0.0s
failed to create LLB definition: the Dockerfile cannot be empty
```

### 2.4 Dockerfile with only `# syntax=`

```bash
docker build -t foo -<<'EOF'
# syntax = docker/dockerfile:1.0.1
EOF
```

```
[+] Building 5.6s (4/4) FINISHED                                                                 
 => [internal] load .dockerignore                                                           0.0s
 => => transferring context: 2B                                                             0.0s
 => [internal] load build definition from Dockerfile                                        0.0s
 => => transferring dockerfile: 72B                                                         0.0s
 => resolve image config for docker.io/docker/dockerfile:1.0.1                              1.9s
 => docker-image://docker.io/docker/dockerfile:1.0.1@sha256:653ad79c9cd2fe8105290b61bc0f26  2.9s
 => => resolve docker.io/docker/dockerfile:1.0.1@sha256:653ad79c9cd2fe8105290b61bc0f261c6f  0.0s
 => => sha256:653ad79c9cd2fe8105290b61bc0f261c6f73a9e9cb1780bdb529c396e04f 2.03kB / 2.03kB  0.0s
 => => sha256:9349a3e94e5a58e7ae582e1a3af64f69cdb1a116cf2a299c24fc4d1b22855f0c 527B / 527B  0.0s
 => => sha256:547e40b4ed88f333655094f16f69c7ba0620c9832035ac164d0e6f4a72905445 635B / 635B  0.0s
 => => sha256:ef68ca38e25b693b8cb72a677c2312c9c9fc312619db12ea4e772307a55d 7.52MB / 7.52MB  2.7s
 => => extracting sha256:ef68ca38e25b693b8cb72a677c2312c9c9fc312619db12ea4e772307a55d3e9b   0.2s
exit code: 2
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Based on the above;

We should return a `ErrInvalidParameter` error, so that a 4xx error is returned by the API https://github.com/moby/moby/blob/d48392a35b157612d28c48eb8e4fad1272fa1442/errdefs/defs.go#L8-L11

This patch would addres the cases I found while testing, but perhaps we need a follow-up to check for other errors;

```patch
diff --git a/builder/remotecontext/detect.go b/builder/remotecontext/detect.go
index 23db8a5aaa..073dee9a6b 100644
--- a/builder/remotecontext/detect.go
+++ b/builder/remotecontext/detect.go
@@ -12,6 +12,7 @@ import (
        "github.com/docker/docker/api/types/backend"
        "github.com/docker/docker/builder"
        "github.com/docker/docker/builder/dockerignore"
+       "github.com/docker/docker/errdefs"
        "github.com/docker/docker/pkg/fileutils"
        "github.com/docker/docker/pkg/urlutil"
        "github.com/moby/buildkit/frontend/dockerfile/parser"
@@ -147,14 +148,14 @@ func readAndParseDockerfile(name string, rc io.Reader) (*parser.Result, error) {
        br := bufio.NewReader(rc)
        if _, err := br.Peek(1); err != nil {
                if err == io.EOF {
-                       return nil, errors.Errorf("the Dockerfile (%s) cannot be empty", name)
+                       return nil, errdefs.InvalidParameter(errors.Errorf("the Dockerfile (%s) cannot be empty", name))
                }
                return nil, errors.Wrap(err, "unexpected error reading Dockerfile")
        }
 
        dockerfile, err := parser.Parse(br)
        if err != nil {
-               return nil, errors.Wrapf(err, "failed to parse %s", name)
+               return nil, errdefs.InvalidParameter(errors.Wrapf(err, "failed to parse %s", name))
        }
 
        return dockerfile, nil
```


In the BuildKit cases;

- (2.1 / 2.2) `failed to create LLB definition: file with no instructions.` is a bit abstract; @tonistiigi - any suggestions for that one?
- (2.3) `failed to create LLB definition: the Dockerfile cannot be empty` - that looks better (clear that it's something to do with the Dockerfile)
- (2.4) `exit code: 2` - this should be improved; probably something to be fixed upstream? @tonistiigi?


Testing:

- Perhaps add a test for 1.2, 2.4

**Comment by LinuxMercedes** [comment]
I have (I think) wrapped the error from `parser.Parse` everywhere it might be raised and added a test for case 1.2.

It seems that case 2.4 is indeed a buildkit problem and should be fixed there to raise the "file with no instructions" error as well. I can put together a patch for that in a bit if y'all agree. Also happy to workshop the wording of that error message since I agree that it's a little abstract for output from `docker build` and the like.

Once I do so I'll bump the vendor version here, add a test for case 2.4, and squash the commits up a bit!

**Comment by olljanat** [comment]
@LinuxMercedes CI build fails to:
```
02:17:33 integration/build/build_test.go:1::warning: file is not gofmted with -s (gofmt)
02:17:33 integration/build/build_test.go:1::warning: file is not goimported (goimports)
02:17:34 Build step 'Execute shell' marked build as failure
```

**Comment by thaJeztah** [comment]
The `gofmt` issue may be due to a different Golang version being used 😞 this is the diff;

```patch
diff --git a/integration/build/build_test.go b/integration/build/build_test.go
index 042f1fa0cb..d79d1f9eb8 100644
--- a/integration/build/build_test.go
+++ b/integration/build/build_test.go
@@ -490,8 +490,8 @@ func TestBuildWithEmptyDockerfile(t *testing.T) {
 			expectedErr: "cannot be empty",
 		},
 		{
-			name:        "empty-lines-dockerfile",
-			dockerfile:  `
+			name: "empty-lines-dockerfile",
+			dockerfile: `
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks for sticking with us 🤗 the changes LGTM, some things remaining;

- fix the `gofmt` to make CI pass 👍 
- could you combine (squash) the `Test for case 1.2` commit with the `Test that dockrefiles with no instructions are detected` commit?
  - oh! while doing that; might want to change the typo in that commit's message (dockrefiles)

Let me know if you need a hand with the squashing/combining (not everyone knows how to 👍)

File: builder/remotecontext/detect.go [hunk 2]
```diff
@@ -146,11 +148,17 @@ func readAndParseDockerfile(name string, rc io.Reader) (*parser.Result, error) {
 	br := bufio.NewReader(rc)
 	if _, err := br.Peek(1); err != nil {
 		if err == io.EOF {
-			return nil, errors.Errorf("the Dockerfile (%s) cannot be empty", name)
+			return nil, errdefs.InvalidParameter(errors.Errorf("the Dockerfile (%s) cannot be empty", name))
 		}
 		return nil, errors.Wrap(err, "unexpected error reading Dockerfile")
```
[the remark this answers is not in the document]
Comment (LinuxMercedes) [reply builder/remotecontext/detect.go]: We **could** just strike that code entirely and let `parser.Parse` deal with it -- as is the case for all of the other codepaths which call `parser.Parse`. If we do keep it, though, I think it makes sense to categorize it the same as a parse error just for that reason.

**Comment by olljanat** [comment]
@LinuxMercedes yes Z failure is this one #38460
but experimental failure I have not seen earlier and I cannot find it from issues:
```
17:19:20 === RUN   TestServiceWithDefaultAddressPoolInit
17:19:21 --- FAIL: TestServiceWithDefaultAddressPoolInit (1.28s)
17:19:21     daemon.go:296: [d38c8a3f9fff9] waiting for daemon to start
17:19:21     daemon.go:328: [d38c8a3f9fff9] daemon started
17:19:21     service_test.go:420: TestServiceWithDefaultAddressPoolInit: NetworkInspect: {Name:saanvisthiraTestServiceWithDefaultAddressPoolInit ID:2e1jfqltui460afgk05cgwr5r Created:2019-01-08 17:19:21.726166875 +0000 UTC Scope:swarm Driver: EnableIPv6:false IPAM:{Driver: Options:map[] Config:[]} Internal:false Attachable:false Ingress:false ConfigFrom:{Network:} ConfigOnly:false Containers:map[] Options:map[] Labels:map[] Peers:[] Services:map[]}
17:19:21     service_test.go:421: assertion failed: expression is false: len(out.IPAM.Config) > 0
```

Need to check if it is related to this one or some recently merged PR.

**Comment by olljanat** [comment]
Derek add label: rebuild/experimental

**Comment by LinuxMercedes** [comment]
Another flaky test in z: https://github.com/moby/moby/issues/37547

**Comment by olljanat** [comment]
> Another flaky test in z and experimental: #37547

Yep. That should be fixed on #38499 hopefully it passes tests and we get it merged soon...

**Comment by LinuxMercedes** [comment]
The experimental failure is indeed a regression -- it expects a 500 from the API whereas it is now a 400 in this PR.

**Comment by thaJeztah** [comment]
> The experimental failure is indeed a regression -- it expects a 500 from the API whereas it is now a 400 in this PR.

Let me check that test; possibly that test is incorrect, and it was testing for the wrong behavior (as in: it happened to return a 500, so the test was written for that)

**Comment by thaJeztah** [comment]
Ah, docker-py test;

```
7:40:17 ________ TestRegressions.test_443_handle_nonchunked_response_in_stream _________
17:40:17 /docker-py/tests/integration/regression_test.py:17: in test_443_handle_nonchunked_response_in_stream
17:40:17     assert exc.value.response.status_code == 500
17:40:17 E   AssertionError: assert 400 == 500
17:40:17 E    +  where 400 = <Response [400]>.status_code
17:40:17 E    +    where <Response [400]> = APIError(HTTPError(u'400 Client Error: Bad Request for url: http+docker://loca.../build?q=False&pull=False&t=a%2Fb%2Fc&nocache=False&forcerm=False&rm=False',),).response
17:40:17 E    +      where APIError(HTTPError(u'400 Client Error: Bad Request for url: http+docker://loca.../build?q=False&pull=False&t=a%2Fb%2Fc&nocache=False&forcerm=False&rm=False',),) = <ExceptionInfo APIError tblen=5>.value
17:40:17 === 1 failed, 327 passed, 5 skipped, 4 xfailed, 1 xpassed in 353.99 seconds ====
```

Looks like it's this test; https://github.com/docker/docker-py/blob/8b246db271a85d6541dc458838627e89c683e42f/tests/integration/regression_test.py#L12-L18

```python
    def test_443_handle_nonchunked_response_in_stream(self):
        dfile = io.BytesIO()
        with pytest.raises(docker.errors.APIError) as exc:
            for line in self.client.build(fileobj=dfile, tag="a/b/c"):
                pass
        assert exc.value.response.status_code == 500
        dfile.close()
```

**Comment by thaJeztah** [comment]
Dug a bit in that repository's history, and I see that test was originally added in https://github.com/docker/docker-py/commit/09ea71971c09837b9fb2c46cb0d8309ff321f16e to address https://github.com/docker/docker-py/issues/443

**Comment by LinuxMercedes** [comment]
I think it'd be reasonable to just fix that test to pass (i.e., s/500/400/). Conversely, that test checks for an empty dockerfile, so we could just not wrap the err [here](https://github.com/moby/moby/blob/ba53e63af5e8cf30b2ceb5cf22f9cff51fefe33a/builder/remotecontext/detect.go#L151) and keep the old behavior.

File: builder/remotecontext/detect.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply builder/remotecontext/detect.go]: I didn't check the whole `br.Peek(1)` code-path; wasn't sure if error could only be due to an error in the _request_ or also originate from something server-side 😅

**Comment by thaJeztah** [comment]
Yes. I'm not really a Python code, and wasn't sure if assert handled `or` there, so here's my quick attempt: https://github.com/docker/docker-py/pull/2216

**Comment by LinuxMercedes** [comment]
z and janky failures are flaky: #37306

Not sure what's going on with the experimental build but it looks unrelated:
```
03:05:02 FAIL: docker_cli_daemon_plugins_test.go:161: DockerDaemonSuite.TestDaemonKillWithPlugins
```
<details>
```
03:05:02 FAIL: docker_cli_daemon_plugins_test.go:161: DockerDaemonSuite.TestDaemonKillWithPlugins
03:05:02 
03:05:02 [db811e51e435f] waiting for daemon to start
03:05:02 [db811e51e435f] daemon started
03:05:02 
03:05:02 docker_cli_daemon_plugins_test.go:166:
03:05:02     c.Fatalf("Could not install plugin: %v %s", err, out)
03:05:02 ... Error: Could not install plugin: exit status 1 latest: Pulling from tiborvass/sample-volume-plugin
03:05:02 eb9c16fbdc53: Verifying Checksum
03:05:02 eb9c16fbdc53: Download complete
03:05:02 error pulling image configuration: unknown blob
03:05:02 
03:05:02 
03:05:02 [db811e51e435f] exiting daemon
03:05:04 time="2019-01-15T03:05:04Z" level=warning msg="Running modprobe nf_nat failed with message: ``, error: exec: \"modprobe\": executable file not found in $PATH"
03:05:04 time="2019-01-15T03:05:04Z" level=warning msg="Running modprobe xt_conntrack failed with message: ``, error: exec: \"modprobe\": executable file not found in $PATH"
```
</details>

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks so much!

**Pull request merged**